### PR TITLE
fix/enhance redirect validation

### PR DESCRIPTION
**Pull request by fiji-flo**

### CODE REVIEW

File: content/redirect.js
Comment (peterbe) [inline content/redirect.js:54]: Shouldn't this be something like
```javascript
if (VALID_LOCALES.map(locale => `/${locale}/`).includes(url)) {
```
Just in case the Japanese wants a particular page to simply go back to the home page?

**Review by peterbe — COMMENTED** [review]
Not done reviewing but I thought I'd drop by a useful comment. 

```
yarn tool add-redirect /en-US/docs/x y
...
fe: The "path" argument must be of type string. Received undefined
```

**Comment by peterbe** [comment]
Which reminds me. We should probably throw an error if someone tries to do:
```
yarn tool add-redirect /en-US/docs/x http://www.nosecure.com
```
But perhaps it's too draconian as we deal with archived junk that needs to go away.

File: content/redirect.js
Comment (fiji-flo) [reply content/redirect.js]: Yes, I'm changing that. In all `_redirects.txt` we only had this for _en-US_

File: content/redirect.js
Comment (fiji-flo) [inline content/redirect.js:197]: I *ignored* this initially because it's a mess. But we should try to introduce strict APIs from where on out URL are _decoded_ and _lowercase_. I hope this doesn't come back at us soon :/

File: tool/cli.js
Comment (fiji-flo) [inline tool/cli.js:57]: Happy to rename. That's what we should run on PRs for `content` and `translated-content`

File: content/redirect.js
Comment (fiji-flo) [inline content/redirect.js:53]: the _en-US_ `_redirects.txt` contains redirects to translated content. In case we don't have that checked out we still want this to work.

File: content/redirect.js
Comment (peterbe) [inline content/redirect.js:23]:
```suggestion
function isVanityRedirectURL(url) {
```
...is the convention. Mentioned in `docs/conventions.md`

Comment (peterbe) [reply content/redirect.js:23]: If you see more of these feel free to "upgrade".

File: content/redirect.js
Comment (peterbe) [inline content/redirect.js:51]:
```suggestion
      `Trying to resolve a non-en-us path for ${url} without CONTENT_TRANSLATED_ROOT set.`
```

File: content/redirect.js
Comment (peterbe) [reply content/redirect.js:196]: I don't get it. Imagine, in the future, where this PR comment isn't around. How do you connect that code comment with the `  const decodedPairs = decodePairs(pairs);` line?

Even with the PR comment I still don't understand the point of the "This is dangerous ..." comment. Sorry. Is calling `decodePairs()` dangerous??

File: tool/cli.js
Comment (peterbe) [inline tool/cli.js:77]: It's so confusing. `validate-redirects` doesn't validate the redirects??
"Verify" is when you check something after you've done something.

File: tool/cli.js
Comment (peterbe) [inline tool/cli.js:72]: supernit: The variable name `l` isn't great because it's hard to read. It looks like a numeral `1`. Also single-letter variables isn't great. `error` is better than `e` and `locale` is better than `l` (even it means you have to do `const locales = args.locale`).

File: tool/cli.js
Comment (peterbe) [inline tool/cli.js:66]: nit
```suggestion
      for (const locale of args.locale) {
```

**Review by peterbe — CHANGES_REQUESTED** [review]
I'm still very confused about the notation of "verify" and "validate". I like that in the content repo it calls `yarn content validate-redirects` and there that naming makes sense. 

By the way, it fails but it looks like it's failing on something unintended:

```
▶ yarn tool add-redirect /en-US/docs/x 11
yarn run v1.22.10
$ node tool/cli.js add-redirect /en-US/docs/x 11

error: url.includes is not a function

```

**Comment by fiji-flo** [comment]
@peterbe 
```
▶ yarn tool add-redirect /en-US/docs/x 11
yarn run v1.22.10
$ node tool/cli.js add-redirect /en-US/docs/x 11

error: url.includes is not a function
```

that happens on master, too. And probably since forever. Caporal casts `11` to a Number and we expect it to be a String.

I've added a global `.cast(false)` to the tool/cli. Mayb we should add this to all caporal clis?

File: tool/cli.js
Comment (fiji-flo) [reply tool/cli.js:78]: https://en.wikipedia.org/wiki/Software_verification_and_validation

I don't know how to name these. I guess right now:
*validate-redirects* asserts that the (all) `_redirects.txt` can be loaded and does it's own basic sanity check.

*verify-redirects* asserts that a (many) `_redirects.txt` have optimized and correct entries by calling `validateTo(From)URL` for each entry, checking encoding and asserts that it's in alphabetical order and optimized.

We should run *validate-redirects* during deployment and yari PRs and *verify-redirects* for content PRs.

Happy to change the name or move to change the `load` function to also _verify_.

File: content/redirect.js
Comment (fiji-flo) [reply content/redirect.js:198]: Maybe I'm just paranoid. We shouldn't need this here as the the `_redirects.txt` should always be decoded since we changed it. We may move to the convention that whatever calls `.add(` does so with _decoded_ pairs. Then we can remove it all together. And maybe have a warning if an internal (beginning with /) URL contains a `%`.

I would do that in a follow up?

**Comment by fiji-flo** [comment]
@peterbe can you give this another look? I hope I addressed the things I could. I still don't know what to do about the naming.
I see 3 options:
1. leave it as it is now
2. come up with better naming
3. combine both (call `Redirect.verifyLocale` from the `validate-redirects` command)

File: content/redirect.js
Comment (peterbe) [reply content/redirect.js:198]: What I mean is that the comment is very hard to read. What's so "dangerous"?
Is it meant to be one of those `// TODO: ...` notes? 
What can go wrong if you call `decodePairs(pairs);` and what effects are there? And if it's bad one, why is it called?

I'm just pointing out that the comment isn't helping. It only makes me nervous without actually understanding the underlying **why**.

Comment (peterbe) [reply content/redirect.js:198]: Beyond the nit about the code comment... You mentioned that decoding should maybe only be done n the `.add()`. Perhaps I don't fully comprehend exactly what's going on but I think if someone ignores the CLI and manually clobbers in some junk into the `_redirects.txt` using the GitHub UI or emacs directly, it must get caught in CI when it eventually becomes a PR. Does that make sense?

File: tool/cli.js
Comment (peterbe) [reply tool/cli.js:78]: From the Wikipedia article:

- `verify` - check that it worked
- `validate` - was it the right check?

I think some guidance is required. The fact that we had to discuss the meaning of the words indicates to me that this is potentially too complicated. (Also, when you skim the code it says that the function is called "verify-redirects" and inside it calls "validateLocale()")

Why does any CLI need to be run "during deployment"?

What we want is two tools:

- One that we can run in content pull requests CI to make sure that the contents of the `_redirects.txt` file isn't corrupt or imperfect. I.e. guarding against people manually messing with the .txt file without. 
- One that we only use very rarely, by the core dev team, to tidy and clean up the `_redirects.txt` file so it's above-and-beyond sane and optimized as some redirects can be short-circuited to perfection. 

Perhaps take one last look at the names and the help strings to see if they can be optimized (for usability/understanding) so they can be easily understood by us in 6 months times (or a new dev joining tomorrow) can understand them without having to break down the code.

**Review by peterbe — COMMENTED** [review]
I've tested it and I think all the core functionality is correct and perfect. I'm still confused as to why you need two different commands and why one of them should be part of deployment. 

If the `validate-redirects` is just a loop over all locales and running `verify-redirects` for each, then I think the `verify-redirects` is uninteresting to have since `validate-redirects` is all you need. (that would have been a different story if it took too long to run each locale so it'd be useful to be able to just run a single locale). 

Lastly, and I don't think this is a point about any particular line of code; but what and how do we periodically run something to see if the file can be optimized more?

File: tool/cli.js
Comment (peterbe) [inline tool/cli.js:55]: I appreciate that the CI needs to change to run `validate-redirects --strict` now. 
But when would you run this CLI *without* the `--strict`?

Comment (peterbe) [reply tool/cli.js:55]: This question is the last time remaining. And once this lands you'll update the `pr-test.yml` in the mdn/content repo, right?

Comment (fiji-flo) [reply tool/cli.js:55]: The idea was to keep the `validate-redirects` for yari CI and use `validate-redirects --strict` for `(translated-)content`

Comment (peterbe) [reply tool/cli.js:55]: When would you run `validate-redirects` in Yari? Just for (manually) sanity checking periodically?

Comment (fiji-flo) [reply tool/cli.js:55]: Keep it like it is. `validate-redirects` still assures that redirects can be loaded.

**Review by peterbe — APPROVED** [review]
For the record, I still don't understand the `--strict` option. (I get that it makes things more strict but I can't understand the context of why you wouldn't always want that)

**Pull request merged**
